### DISCUSSION TITLE
Key value pair JSON object as a field (using GraphQLJSON)

### DISCUSSION THREAD

**Discussion by tusharchutani** [MichalLytek/type-graphql#823] [General]
Hey, 

I have a requirement for a generic type for an input (they are more specifically filters that get passed in. For brevity I have I have omiited out a big chunk of code. Here is the input type. 

```
@InputType()
export class GetFilterValueArgs {

    @Field(type => ID)
    id: string;

    @Field(type => GraphQLJSON, { defaultValue: null })
    appliedFilters?: { [key: string]: Filter; }
...
```
The filter type looks like this:
```
@InputType()
class Filter {
    @Field(type => [String])
    values: string[]

    @Field({ defaultValue: false })
    exclude: boolean
}
```
Reading other forms on this topic I thought the issue would be the import but I have both `allowSyntheticDefaultImports` and `esModuleInterop` enabled. I am importing GraphQLJSON like `import * as GraphQLJSON from 'graphql-type-json';` 

I am getting an unhandled promise issue like this 
```
UnhandledPromiseRejectionWarning: Error: Cannot determine GraphQL input type for 'appliedFilters' of 'GetFilterValueArgs' class.
```
Some help on this topic would be helpful

**Comment by MichalLytek** [comment]
As you have "have both allowSyntheticDefaultImports and esModuleInterop enabled", the proper import should look like:
```ts
import GraphQLJSON from 'graphql-type-json';
```

**Reply by tusharchutani** [reply]
ah yes sorry you are right. I think I totally misunderstood the meaning of these options. I am not sure if this possible but is it possible for us to have a key value pair field for our input type. Right now (after fixing the import of GraphQLJson) the type of the argument is coming out to be JSON . I am using apollo codegen for our react app to generate types and I would like to go the route of having a type of appliedFilter defined as `{ [key: string]: Filter; }`.

**Reply by MichalLytek** [reply]
JSON is JSON, it can contain anything 😛 

Maybe you can return an array of `{ key: string; value: Filter }`? So you could model this as `Object.entries` with type-safety (Record of type Filter).

You could even go further and have an enum describing all the keys of some type 😉

**Reply by tusharchutani** [reply]
ahh see yeah thanks that was hunch on this. In our usecase our keys are unknown before hand so I think going this route would probably be the best. Thank you so much
